File: pkg/client/operations_test.go
package client_test

import (
	"bytes"
	"context"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"

	"github.com/candlekeep/zot-artifact-store/pkg/client"
	"github.com/candlekeep/zot-artifact-store/test"
)

func TestBucketOperations(t *testing.T) {
	t.Run("Create bucket", func(t *testing.T) {
		// Given: Test server
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			test.AssertEqual(t, "PUT", r.Method, "HTTP method")
			test.AssertEqual(t, "/s3/test-bucket", r.URL.Path, "request path")
			w.WriteHeader(http.StatusOK)
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Creating a bucket
		ctx := context.Background()
		err := c.CreateBucket(ctx, "test-bucket")

		// Then: Request succeeds
		test.AssertNoError(t, err, "create bucket")
	})

	t.Run("Delete bucket", func(t *testing.T) {
		// Given: Test server
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			test.AssertEqual(t, "DELETE", r.Method, "HTTP method")
			test.AssertEqual(t, "/s3/test-bucket", r.URL.Path, "request path")
			w.WriteHeader(http.StatusOK)
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Deleting a bucket
		ctx := context.Background()
		err := c.DeleteBucket(ctx, "test-bucket")

		// Then: Request succeeds
		test.AssertNoError(t, err, "delete bucket")
	})

	t.Run("List buckets", func(t *testing.T) {
		// Given: Test server returning buckets
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			test.AssertEqual(t, "GET", r.Method, "HTTP method")
			test.AssertEqual(t, "/s3", r.URL.Path, "request path")
			w.WriteHeader(http.StatusOK)
			w.Write([]byte(`{"buckets": [{"name": "bucket1", "creationDate": "2024-01-01T00:00:00Z"}]}`))
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Listing buckets
		ctx := context.Background()
		result, err := c.ListBuckets(ctx)

		// Then: Returns bucket list
		test.AssertNoError(t, err, "list buckets")
		test.AssertTrue(t, result != nil, "result should not be nil")
		test.AssertTrue(t, len(result.Buckets) == 1, "should have 1 bucket")
		test.AssertEqual(t, "bucket1", result.Buckets[0].Name, "bucket name")
	})
}

func TestObjectOperations(t *testing.T) {
	t.Run("Upload object", func(t *testing.T) {
		// Given: Test server and upload data
		uploadData := []byte("test artifact content")
		var receivedData []byte
		var receivedContentType string

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			test.AssertEqual(t, "PUT", r.Method, "HTTP method")
			test.AssertEqual(t, "/s3/test-bucket/test-key", r.URL.Path, "request path")
			receivedContentType = r.Header.Get("Content-Type")
			receivedData, _ = io.ReadAll(r.Body)
			w.WriteHeader(http.StatusOK)
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Uploading an object
		ctx := context.Background()
		opts := &client.UploadOptions{
			ContentType: "application/gzip",
		}
		err := c.Upload(ctx, "test-bucket", "test-key", bytes.NewReader(uploadData), int64(len(uploadData)), opts)

		// Then: Upload succeeds with correct data
		test.AssertNoError(t, err, "upload")
		test.AssertEqual(t, "application/gzip", receivedContentType, "content type")
		test.AssertEqual(t, string(uploadData), string(receivedData), "uploaded data")
	})

	t.Run("Upload with metadata", func(t *testing.T) {
		// Given: Test server and upload data with metadata
		uploadData := []byte("test content")
		var receivedMetadata map[string]string

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			receivedMetadata = make(map[string]string)
			for key, values := range r.Header {
				if strings.HasPrefix(key, "X-Amz-Meta-") {
					metaKey := key[11:]
					if len(values) > 0 {
						receivedMetadata[metaKey] = values[0]
					}
				}
			}
			w.WriteHeader(http.StatusOK)
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Uploading with metadata
		ctx := context.Background()
		opts := &client.UploadOptions{
			Metadata: map[string]string{
				"version": "1.0",
				"author":  "test-user",
			},
		}
		err := c.Upload(ctx, "test-bucket", "test-key", bytes.NewReader(uploadData), int64(len(uploadData)), opts)

		// Then: Metadata is included in request
		test.AssertNoError(t, err, "upload")
		test.AssertEqual(t, "1.0", receivedMetadata["Version"], "version metadata")
		test.AssertEqual(t, "test-user", receivedMetadata["Author"], "author metadata")
	})

	t.Run("Download object", func(t *testing.T) {
		// Given: Test server with artifact data
		artifactData := []byte("downloaded artifact content")

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			test.AssertEqual(t, "GET", r.Method, "HTTP method")
			test.AssertEqual(t, "/s3/test-bucket/test-key", r.URL.Path, "request path")
			w.WriteHeader(http.StatusOK)
			w.Write(artifactData)
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Downloading an object
		ctx := context.Background()
		var buffer bytes.Buffer
		err := c.Download(ctx, "test-bucket", "test-key", &buffer, nil)

		// Then: Download succeeds with correct data
		test.AssertNoError(t, err, "download")
		test.AssertEqual(t, string(artifactData), buffer.String(), "downloaded data")
	})

	t.Run("Download with range", func(t *testing.T) {
		// Given: Test server with range support
		var receivedRange string

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			receivedRange = r.Header.Get("Range")
			w.WriteHeader(http.StatusPartialContent)
			w.Write([]byte("partial"))
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Downloading with range
		ctx := context.Background()
		var buffer bytes.Buffer
		opts := &client.DownloadOptions{
			Range: "bytes=0-1023",
		}
		err := c.Download(ctx, "test-bucket", "test-key", &buffer, opts)

		// Then: Range header is set
		test.AssertNoError(t, err, "download")
		test.AssertEqual(t, "bytes=0-1023", receivedRange, "range header")
	})

	t.Run("Get object metadata", func(t *testing.T) {
		// Given: Test server with object metadata
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			test.AssertEqual(t, "HEAD", r.Method, "HTTP method")
			w.Header().Set("Content-Length", "1024")
			w.Header().Set("Content-Type", "application/gzip")
			w.Header().Set("ETag", "abc123")
			w.Header().Set("X-Amz-Meta-version", "1.0")
			w.WriteHeader(http.StatusOK)
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Getting object metadata
		ctx := context.Background()
		obj, err := c.GetObjectMetadata(ctx, "test-bucket", "test-key")

		// Then: Metadata is returned
		test.AssertNoError(t, err, "get metadata")
		test.AssertTrue(t, obj != nil, "object should not be nil")
		test.AssertEqual(t, int64(1024), obj.Size, "object size")
		test.AssertEqual(t, "application/gzip", obj.ContentType, "content type")
		test.AssertEqual(t, "abc123", obj.ETag, "etag")
		test.AssertEqual(t, "1.0", obj.Metadata["Version"], "custom metadata")
	})

	t.Run("Delete object", func(t *testing.T) {
		// Given: Test server
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			test.AssertEqual(t, "DELETE", r.Method, "HTTP method")
			test.AssertEqual(t, "/s3/test-bucket/test-key", r.URL.Path, "request path")
			w.WriteHeader(http.StatusOK)
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Deleting an object
		ctx := context.Background()
		err := c.DeleteObject(ctx, "test-bucket", "test-key")

		// Then: Delete succeeds
		test.AssertNoError(t, err, "delete object")
	})

	t.Run("List objects", func(t *testing.T) {
		// Given: Test server returning objects
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			test.AssertEqual(t, "GET", r.Method, "HTTP method")
			test.AssertTrue(t, strings.Contains(r.URL.Path, "/s3/test-bucket"), "request path")
			w.WriteHeader(http.StatusOK)
			w.Write([]byte(`{"contents": [{"key": "obj1", "size": 100}], "isTruncated": false}`))
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Listing objects
		ctx := context.Background()
		result, err := c.ListObjects(ctx, "test-bucket", nil)

		// Then: Returns object list
		test.AssertNoError(t, err, "list objects")
		test.AssertTrue(t, result != nil, "result should not be nil")
		test.AssertTrue(t, len(result.Objects) == 1, "should have 1 object")
		test.AssertEqual(t, "obj1", result.Objects[0].Key, "object key")
		test.AssertEqual(t, int64(100), result.Objects[0].Size, "object size")
	})

	t.Run("List objects with prefix", func(t *testing.T) {
		// Given: Test server
		var receivedPrefix string

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			receivedPrefix = r.URL.Query().Get("prefix")
			w.WriteHeader(http.StatusOK)
			w.Write([]byte(`{"contents": [], "isTruncated": false}`))
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Listing with prefix
		ctx := context.Background()
		opts := &client.ListOptions{
			Prefix: "test/",
		}
		c.ListObjects(ctx, "test-bucket", opts)

		// Then: Prefix is included in query
		test.AssertEqual(t, "test/", receivedPrefix, "prefix parameter")
	})

	t.Run("Copy object", func(t *testing.T) {
		// Given: Test server
		var receivedCopySource string

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			test.AssertEqual(t, "PUT", r.Method, "HTTP method")
			receivedCopySource = r.Header.Get("X-Amz-Copy-Source")
			w.WriteHeader(http.StatusOK)
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Copying an object
		ctx := context.Background()
		err := c.CopyObject(ctx, "source-bucket", "source-key", "dest-bucket", "dest-key")

		// Then: Copy succeeds with correct headers
		test.AssertNoError(t, err, "copy object")
		test.AssertEqual(t, "/source-bucket/source-key", receivedCopySource, "copy source header")
	})
}

func TestMultipartUpload(t *testing.T) {
	t.Run("Initiate multipart upload", func(t *testing.T) {
		// Given: Test server
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			test.AssertEqual(t, "POST", r.Method, "HTTP method")
			test.AssertTrue(t, strings.Contains(r.URL.RawQuery, "uploads"), "should have uploads parameter")
			w.WriteHeader(http.StatusOK)
			w.Write([]byte(`{"uploadId": "test-upload-id"}`))
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Initiating multipart upload
		ctx := context.Background()
		upload, err := c.InitiateMultipartUpload(ctx, "test-bucket", "test-key", nil)

		// Then: Multipart upload is initiated
		test.AssertNoError(t, err, "initiate multipart upload")
		test.AssertTrue(t, upload != nil, "upload should not be nil")
		test.AssertEqual(t, "test-upload-id", upload.UploadID, "upload ID")
		test.AssertEqual(t, "test-bucket", upload.Bucket, "bucket")
		test.AssertEqual(t, "test-key", upload.Key, "key")
	})

	t.Run("Upload part", func(t *testing.T) {
		// Given: Test server and multipart upload
		partData := []byte("part content")

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if strings.Contains(r.URL.RawQuery, "uploads") {
				w.WriteHeader(http.StatusOK)
				w.Write([]byte(`{"uploadId": "test-upload-id"}`))
			} else {
				test.AssertTrue(t, strings.Contains(r.URL.RawQuery, "uploadId=test-upload-id"), "should have uploadId")
				test.AssertTrue(t, strings.Contains(r.URL.RawQuery, "partNumber=1"), "should have partNumber")
				w.Header().Set("ETag", "part-etag-1")
				w.WriteHeader(http.StatusOK)
			}
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})
		ctx := context.Background()
		upload, _ := c.InitiateMultipartUpload(ctx, "test-bucket", "test-key", nil)

		// When: Uploading a part
		etag, err := upload.UploadPart(ctx, 1, bytes.NewReader(partData), int64(len(partData)))

		// Then: Part upload succeeds
		test.AssertNoError(t, err, "upload part")
		test.AssertEqual(t, "part-etag-1", etag, "part etag")
	})

	t.Run("Complete multipart upload", func(t *testing.T) {
		// Given: Test server and multipart upload
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if strings.Contains(r.URL.RawQuery, "uploads") {
				w.WriteHeader(http.StatusOK)
				w.Write([]byte(`{"uploadId": "test-upload-id"}`))
			} else if r.Method == "POST" {
				test.AssertTrue(t, strings.Contains(r.URL.RawQuery, "uploadId=test-upload-id"), "should have uploadId")
				w.WriteHeader(http.StatusOK)
			}
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})
		ctx := context.Background()
		upload, _ := c.InitiateMultipartUpload(ctx, "test-bucket", "test-key", nil)

		// When: Completing multipart upload
		parts := []client.CompletedPart{
			{PartNumber: 1, ETag: "etag1"},
			{PartNumber: 2, ETag: "etag2"},
		}
		err := upload.Complete(ctx, parts)

		// Then: Multipart upload completes
		test.AssertNoError(t, err, "complete multipart upload")
	})

	t.Run("Abort multipart upload", func(t *testing.T) {
		// Given: Test server and multipart upload
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if strings.Contains(r.URL.RawQuery, "uploads") && r.Method == "POST" {
				w.WriteHeader(http.StatusOK)
				w.Write([]byte(`{"uploadId": "test-upload-id"}`))
			} else if r.Method == "DELETE" {
				test.AssertTrue(t, strings.Contains(r.URL.RawQuery, "uploadId=test-upload-id"), "should have uploadId")
				w.WriteHeader(http.StatusOK)
			}
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})
		ctx := context.Background()
		upload, _ := c.InitiateMultipartUpload(ctx, "test-bucket", "test-key", nil)

		// When: Aborting multipart upload
		err := upload.Abort(ctx)

		// Then: Multipart upload is aborted
		test.AssertNoError(t, err, "abort multipart upload")
	})
}

func TestProgressCallbacks(t *testing.T) {
	t.Run("Upload progress callback", func(t *testing.T) {
		// Given: Test server and progress tracking
		uploadData := []byte("test content for progress")
		var progressCalled bool
		var totalBytes int64

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			io.ReadAll(r.Body)
			w.WriteHeader(http.StatusOK)
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Uploading with progress callback
		ctx := context.Background()
		opts := &client.UploadOptions{
			ProgressCallback: func(bytes int64) {
				progressCalled = true
				totalBytes = bytes
			},
		}
		err := c.Upload(ctx, "test-bucket", "test-key", bytes.NewReader(uploadData), int64(len(uploadData)), opts)

		// Then: Progress callback is called
		test.AssertNoError(t, err, "upload")
		test.AssertTrue(t, progressCalled, "progress callback should be called")
		test.AssertTrue(t, totalBytes > 0, "total bytes should be tracked")
	})

	t.Run("Download progress callback", func(t *testing.T) {
		// Given: Test server and progress tracking
		downloadData := []byte("test download content")
		var progressCalled bool
		var totalBytes int64

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.WriteHeader(http.StatusOK)
			w.Write(downloadData)
		}))
		defer server.Close()

		c, _ := client.NewClient(&client.Config{BaseURL: server.URL})

		// When: Downloading with progress callback
		ctx := context.Background()
		var buffer bytes.Buffer
		opts := &client.DownloadOptions{
			ProgressCallback: func(bytes int64) {
				progressCalled = true
				totalBytes = bytes
			},
		}
		err := c.Download(ctx, "test-bucket", "test-key", &buffer, opts)

		// Then: Progress callback is called
		test.AssertNoError(t, err, "download")
		test.AssertTrue(t, progressCalled, "progress callback should be called")
		test.AssertTrue(t, totalBytes > 0, "total bytes should be tracked")
	})
}

File: pkg/client-js/tests/operations.test.ts
/**
 * Tests for operations
 */

import { Client } from '../src/client';
import { Config } from '../src/types';
import axios from 'axios';

// Mock axios
jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe('Operations', () => {
  let client: Client;
  let mockAxiosInstance: any;

  beforeEach(() => {
    jest.clearAllMocks();

    mockAxiosInstance = {
      get: jest.fn(),
      post: jest.fn(),
      put: jest.fn(),
      delete: jest.fn(),
      head: jest.fn(),
      defaults: {
        headers: {
          common: {},
        },
      },
      interceptors: {
        response: {
          use: jest.fn(),
        },
      },
    };

    mockedAxios.create.mockReturnValue(mockAxiosInstance);

    const config: Config = {
      baseURL: 'https://test.example.com',
      token: 'test-token',
    };
    client = new Client(config);
  });

  describe('Bucket Operations', () => {
    it('should create bucket', async () => {
      // Given: Bucket name
      mockAxiosInstance.put.mockResolvedValue({ data: {} });

      // When: Creating bucket
      await client.createBucket('mybucket');

      // Then: Should call PUT /s3/mybucket
      expect(mockAxiosInstance.put).toHaveBeenCalledWith('/s3/mybucket');
    });

    it('should delete bucket', async () => {
      // Given: Existing bucket
      mockAxiosInstance.delete.mockResolvedValue({ data: {} });

      // When: Deleting bucket
      await client.deleteBucket('mybucket');

      // Then: Should call DELETE /s3/mybucket
      expect(mockAxiosInstance.delete).toHaveBeenCalledWith('/s3/mybucket');
    });

    it('should list buckets', async () => {
      // Given: Multiple buckets exist
      mockAxiosInstance.get.mockResolvedValue({
        data: {
          buckets: [
            { name: 'bucket1', creationDate: '2024-01-15T10:30:00Z' },
            { name: 'bucket2', creationDate: '2024-01-16T14:20:00Z' },
          ],
        },
      });

      // When: Listing buckets
      const result = await client.listBuckets();

      // Then: Should return all buckets
      expect(result.buckets).toHaveLength(2);
      expect(result.buckets[0].name).toBe('bucket1');
      expect(result.buckets[1].name).toBe('bucket2');
      expect(result.buckets[0].creationDate).toBeInstanceOf(Date);
    });
  });

  describe('Object Operations', () => {
    it('should upload object', async () => {
      // Given: Binary data to upload
      mockAxiosInstance.put.mockResolvedValue({ data: {} });
      const data = Buffer.from('test data');

      // When: Uploading artifact
      await client.upload('mybucket', 'myfile.tar.gz', data);

      // Then: Should call PUT with data
      expect(mockAxiosInstance.put).toHaveBeenCalledWith(
        '/s3/mybucket/myfile.tar.gz',
        data,
        expect.any(Object)
      );
    });

    it('should upload with metadata', async () => {
      // Given: Artifact with metadata
      mockAxiosInstance.put.mockResolvedValue({ data: {} });
      const data = Buffer.from('test data');

      // When: Uploading with metadata
      await client.upload('mybucket', 'myfile.tar.gz', data, {
        metadata: { version: '1.0.0', author: 'test' },
      });

      // Then: Should include metadata headers
      expect(mockAxiosInstance.put).toHaveBeenCalledWith(
        '/s3/mybucket/myfile.tar.gz',
        data,
        expect.objectContaining({
          headers: expect.objectContaining({
            'X-Amz-Meta-version': '1.0.0',
            'X-Amz-Meta-author': 'test',
          }),
        })
      );
    });

    it('should download object', async () => {
      // Given: Artifact exists
      mockAxiosInstance.get.mockResolvedValue({
        data: Buffer.from('downloaded data'),
      });

      // When: Downloading artifact
      const result = await client.download('mybucket', 'myfile.tar.gz');

      // Then: Should return data
      expect(result).toBeInstanceOf(Buffer);
      expect(result.toString()).toBe('downloaded data');
    });

    it('should download with range', async () => {
      // Given: Artifact exists
      mockAxiosInstance.get.mockResolvedValue({
        data: Buffer.from('partial'),
      });

      // When: Downloading with range
      await client.download('mybucket', 'myfile.tar.gz', {
        range: 'bytes=0-6',
      });

      // Then: Should include Range header
      expect(mockAxiosInstance.get).toHaveBeenCalledWith(
        '/s3/mybucket/myfile.tar.gz',
        expect.objectContaining({
          headers: expect.objectContaining({
            Range: 'bytes=0-6',
          }),
        })
      );
    });

    it('should get object metadata', async () => {
      // Given: Artifact exists
      mockAxiosInstance.head.mockResolvedValue({
        headers: {
          'content-length': '1024',
          'content-type': 'application/gzip',
          etag: '"abc123"',
          'last-modified': 'Mon, 15 Jan 2024 10:30:00 GMT',
          'x-amz-meta-version': '1.0.0',
        },
      });

      // When: Getting metadata
      const obj = await client.getObjectMetadata('mybucket', 'myfile.tar.gz');

      // Then: Should return metadata
      expect(obj.key).toBe('myfile.tar.gz');
      expect(obj.size).toBe(1024);
      expect(obj.contentType).toBe('application/gzip');
      expect(obj.etag).toBe('abc123');
      expect(obj.metadata?.version).toBe('1.0.0');
    });

    it('should delete object', async () => {
      // Given: Artifact exists
      mockAxiosInstance.delete.mockResolvedValue({ data: {} });

      // When: Deleting artifact
      await client.deleteObject('mybucket', 'myfile.tar.gz');

      // Then: Should call DELETE
      expect(mockAxiosInstance.delete).toHaveBeenCalledWith('/s3/mybucket/myfile.tar.gz');
    });

    it('should list objects', async () => {
      // Given: Objects in bucket
      mockAxiosInstance.get.mockResolvedValue({
        data: {
          contents: [
            {
              key: 'file1.tar.gz',
              size: 1024,
              lastModified: '2024-01-15T10:30:00Z',
              etag: 'abc',
              contentType: 'application/gzip',
            },
            {
              key: 'file2.tar.gz',
              size: 2048,
              lastModified: '2024-01-16T14:20:00Z',
              etag: 'def',
              contentType: 'application/gzip',
            },
          ],
          prefix: '',
          maxKeys: 1000,
          isTruncated: false,
        },
      });

      // When: Listing objects
      const result = await client.listObjects('mybucket');

      // Then: Should return objects
      expect(result.objects).toHaveLength(2);
      expect(result.objects[0].key).toBe('file1.tar.gz');
      expect(result.objects[0].size).toBe(1024);
      expect(result.objects[1].key).toBe('file2.tar.gz');
    });

    it('should list objects with prefix', async () => {
      // Given: Objects with prefix
      mockAxiosInstance.get.mockResolvedValue({
        data: {
          contents: [],
          prefix: 'app/',
          maxKeys: 1000,
          isTruncated: false,
        },
      });

      // When: Listing with prefix
      const result = await client.listObjects('mybucket', { prefix: 'app/' });

      // Then: Should include prefix in params
      expect(mockAxiosInstance.get).toHaveBeenCalledWith(
        '/s3/mybucket',
        expect.objectContaining({
          params: expect.objectContaining({
            prefix: 'app/',
          }),
        })
      );
      expect(result.prefix).toBe('app/');
    });

    it('should copy object', async () => {
      // Given: Source artifact exists
      mockAxiosInstance.put.mockResolvedValue({ data: {} });

      // When: Copying artifact
      await client.copyObject('srcbucket', 'srcfile.tar.gz', 'destbucket', 'destfile.tar.gz');

      // Then: Should include copy source header
      expect(mockAxiosInstance.put).toHaveBeenCalledWith(
        '/s3/destbucket/destfile.tar.gz',
        null,
        expect.objectContaining({
          headers: expect.objectContaining({
            'X-Amz-Copy-Source': '/srcbucket/srcfile.tar.gz',
          }),
        })
      );
    });
  });

  describe('Multipart Upload', () => {
    it('should initiate multipart upload', async () => {
      // Given: Large file to upload
      mockAxiosInstance.post.mockResolvedValue({
        data: { uploadId: 'upload123' },
      });

      // When: Initiating multipart upload
      const upload = await client.initiateMultipartUpload('mybucket', 'largefile.tar.gz');

      // Then: Should return upload ID
      expect(upload.uploadId).toBe('upload123');
      expect(upload.bucket).toBe('mybucket');
      expect(upload.key).toBe('largefile.tar.gz');
    });

    it('should upload part', async () => {
      // Given: Initiated multipart upload
      mockAxiosInstance.put.mockResolvedValue({
        headers: { etag: '"part-etag-1"' },
      });

      const data = Buffer.from('part data');

      // When: Uploading part
      const etag = await client.uploadPart('mybucket', 'largefile.tar.gz', 'upload123', 1, data);

      // Then: Should return ETag
      expect(etag).toBe('part-etag-1');
    });

    it('should complete multipart upload', async () => {
      // Given: All parts uploaded
      mockAxiosInstance.post.mockResolvedValue({ data: {} });

      const parts = [
        { partNumber: 1, etag: 'etag1' },
        { partNumber: 2, etag: 'etag2' },
      ];

      // When: Completing upload
      await client.completeMultipartUpload('mybucket', 'largefile.tar.gz', 'upload123', parts);

      // Then: Should send parts list
      expect(mockAxiosInstance.post).toHaveBeenCalledWith(
        '/s3/mybucket/largefile.tar.gz',
        expect.objectContaining({
          parts: expect.arrayContaining([
            { partNumber: 1, etag: 'etag1' },
            { partNumber: 2, etag: 'etag2' },
          ]),
        }),
        expect.objectContaining({
          params: { uploadId: 'upload123' },
        })
      );
    });

    it('should abort multipart upload', async () => {
      // Given: Initiated multipart upload
      mockAxiosInstance.delete.mockResolvedValue({ data: {} });

      // When: Aborting upload
      await client.abortMultipartUpload('mybucket', 'largefile.tar.gz', 'upload123');

      // Then: Should call DELETE
      expect(mockAxiosInstance.delete).toHaveBeenCalledWith(
        '/s3/mybucket/largefile.tar.gz',
        expect.objectContaining({
          params: { uploadId: 'upload123' },
        })
      );
    });
  });
});
